packages/ref-core/src/ref_core/constraints.py

```diff
@@ -0,0 +1,148 @@
+    """
+
+    def apply(self, group: pd.DataFrame, data_catalog: pd.DataFrame) -> pd.DataFrame:
+        """
+        Perform an operation on the group of datasets.
+
+        If this operation modifies the group, the modified group should be returned.
+        If this operation fails, a ConstraintNotSatisfied exception should be raised.
+
+        Parameters
+        ----------
+        group
+            A group of datasets that is being validated.
+        data_catalog
+            The data catalog of datasets
+
+        Raises
+        ------
+        ConstraintNotSatisfied
+            The operation was not successful
+
+        Returns
+        -------
+        :
+            The updated group of datasets
+        """
+        ...
+
+
+Constraint = GroupOperation | GroupValidator
+"""
+A constraint that must be satisfied when executing a given metric run.
+
+This can include operations that are applied to a group of datasets which may modify the group,
+but may also include validators that check if the group satisfies a certain condition.
+
+All constraints must be satisfied for a given group to be run.
+If any constraint is not satisfied, the group will not be executed.
+"""
+
+
+def apply_constraint(
+    dataframe: pd.DataFrame, constraint: Constraint, data_catalog: pd.DataFrame
+) -> pd.DataFrame | None:
+    """
+    Apply a constraint to a group of datasets
+
+    Parameters
+    ----------
+    constraint
+        The constraint to apply
+    data_catalog
+        The data catalog of datasets
+
+    Returns
+    -------
+    :
+        The updated group of datasets or None if the constraint was not satisfied
+    """
+    try:
+        updated_group = (
+            constraint.apply(dataframe, data_catalog) if isinstance(constraint, GroupOperation) else dataframe
+        )
+
+        valid = constraint.validate(updated_group) if isinstance(constraint, GroupValidator) else True
+        if not valid:
+            logger.debug(f"Constraint {constraint} not satisfied for {dataframe}")
+            raise ConstraintNotSatisfied(f"Constraint {constraint} not satisfied for {dataframe}")
+    except ConstraintNotSatisfied:
+        logger.debug(f"Constraint {constraint} not satisfied for {dataframe}")
+        return None
+
+    return updated_group
+
+
+@frozen
+class RequiredFacets:
+    """
+    A constraint that requires a dataset to have certain facets.
+    """
+
+    dimension: str
+    required_facets: list[str]
+
+    def validate(self, group: pd.DataFrame) -> bool:
+        """
+        Check that the required facets are present in the group
+        """
+        if self.dimension not in group:
+            logger.warning(f"Dimension {self.dimension} not present in group {group}")
+            return False
+        return all(value in group[self.dimension].values for value in self.required_facets)
+
+
+@frozen
+class SelectParentExperiment:
+    """
+    Include a dataset's parent experiment in the selection
+    """
```

Comment:
Oh, so the Group of datasets returned from a GroupOperation does not need to be a subset of the input Group of datasets? Fancy. So it could also contain datasets from the data_catalog?

Reply:
This raises the question what happens with multiple Constraints. Do they need to be ordered, so that the GroupOperations are applied in order?

Reply:
> So it could also contain datasets from the data_catalog?

Yes. We need a lot of flexibility here

> Do they need to be ordered, so that the GroupOperations are applied in order?
They are ordered and are applied in that same order. If any fail then the group is skipped.

I think I documented that, but I'll double check